### ISSUE TITLE
Flairs

### ISSUE THREAD

**Issue by regakakobigman** [krisives/cronnit.us#2]
I'd love to use this for automated posting, but the subreddit I'd like to post in uses flairs to categorize submissions. It would be great if you could set a flair to be used for each post.

**Comment by Jurryaany** [comment]
I'll gladly donate in order to give this rather simple feature priority :)

**Comment by krisives** [comment]
Thanks for reminding me to work on this. I'll go ahead and document the current status on this here while I am at it. From what I can tell this can be done by adding the `flair_id` or `flair_text` to the `/api/submit` call to the Reddit API.

For this to work I basically need to modify the Cronnit user interface to let them select which flair they want, but to do that I need to wait for the user to submit the subreddit so I can fetch the flairs from Reddit itself. Unless it's possible for the `flair_id` to simply be a string like "Announcements" which I don't think it is.

**Comment by Jurryaany** [comment]
I understand, it is more complex than it appears, I thought about this a little myself as well after commenting.

What also complicates matters more is that not all subreddits treat flair in the same way, and you can do more or less with it depending on whether you have moderator rights or not on that subreddit.

When you refer to flair_id, is that the same as the css class that the subreddit flair settings include? It's not possible for regular users to view those, but on moderated subreddits it should be visible under /r/subredditname/about/flair/#link_templates.

I think that at the very least it'd be possible to have a field where a user can enter the flair_id (assuming it's the same as that flair's CSS class or default text) and have that applied to the post after it has been submitted.

What are your thoughts?

**Comment by krisives** [comment]
> When you refer to flair_id, is that the same as the css class that the subreddit flair settings include?

I think for `old.reddit.com`, they do use the `flair_id` as part of the CSS. I actually browse Reddit with specialized CSS disabled since many subreddits often make poor choices related to theming such as bad color contrast or hiding downvote buttons, etc.

Also worth mentioning is that old Reddit doesn't actually let you submit a flair until *after* your post has gone live, whereas redesign/new Reddit does let you pick a flair before submitting.

Confusing things a bit more is that the default sub-reddit I use for testing, `/r/test` doesn't seem to have flairs. Doing a bit of digging I was able to figure out that you can list all flairs for a sub-reddit using something like:

    https://www.reddit.com/r/btc/api/link_flair.json
 
There is also `link_flair_v2`, but as you mentioned these are API endpoints that have to be accessed per-user. That's not much harder to do on Cronnit because we have OAuth tokens and can make API requests on behalf of the user.

Current plan of action to do this basically requires adding a few pieces:

* A server side script that will return a list of flairs for a given sub-reddit. It's easiest to do it this way because otherwise taking the OAuth token and giving it to the browser to make requests is more complicated and could be a security concern. (Right now if there is a problem with a token we know *we* caused that, whereas if we give the token to a client and they misuse it, who knows)

* Add an Ajax request to the Submit/Edit form so that it queries the new script asking for a list of flair for the given subreddit selected. This could get confusing if I ever plan to add comma-separated subreddits like I had planned, but we'll cross that bridge when we get there.

* On save the flair ID to be used to the database to a column like `flair`.

* During submission add the `flair_id` if present in the database.

* Eventually adding support to the Batch Import feature so that a column can specify text names of Flair and it can fetch those before doing the import would be useful, but we'll cross that bridge at a later time.

**Comment by Jurryaany** [comment]
Impressive, this is why I'm in the networking part of IT and not programming!

Best of luck, I really hope we can get this to work with the batch import as well at some point.

I'll donate to speed you along :)

**Comment by matthewstrasiotto** [comment]
(Migrated from #63 )

Some subreddits have a requirement that posts be flaired in order to submit.

Reddit exposes both an API for querying usable fairs: 

https://www.reddit.com/r/personalfinance/api/link_flair_v2.json

https://www.reddit.com/dev/api/#GET_api_link_flair_v2

**Comment by Iginis** [comment]
Would love to see this in live.

**Comment by JacobKowalski** [comment]
Any possibility that this could get implemented? Just the .csv support would be fine where we could manually check the flair necessary to post. I would be happy to donate if this is implemented, as I'm running my own instance :)

thanks for the great code!

**Comment by krisives** [comment]
Sure. Just the basic support for `flair_id` and `flair_text` in the CSV should be pretty easy to do.

**Comment by Convoy20** [comment]
@krisives Did the flair_id / flair_text rows get added to the CSV? Happy to update website documentation to reflect it, I've had a couple users request a scheduler with flair support on a subreddit I run

**Comment by Convoy20** [comment]
@krisives I tried to add this in, I don't have a test environment up so I couldn't test it: https://github.com/krisives/cronnit.us/pull/82

**Comment by matthewstrasiotto** [comment]
@Convoy20 there's a vagrantfile & a dockerfile for testing

**Comment by Convoy20** [comment]
> @Convoy20 there's a vagrantfile & a dockerfile for testing

Haven't used either in my life before >_>

**Comment by cesarmonsalud** [comment]
Hello, so can we add flairs yet?

**Comment by krisives** [comment]
Some basic flair support was added in #82. Thanks @Convoy20

**Comment by DaneEvans** [comment]
@krisives It doesn't appear to be on the web version. Just checking that's intentional. 

I took the column names from the code, so it's not just a discrepancy there, and the rest of the uploading of the .csv seems to work. 

If it's intentional, I may just have to figure out how to host my own that's running post #82

**Comment by matthewstrasiotto** [comment]
@DaneEvans , I believe that backend support for flairs was added to the web version, but hasn't been added to the frontend yet. 
I can probably implement that and put in a PR

**Comment by krisives** [comment]
Sorry I have fixed that now and it should be live on the site. There was a problem where a merge conflict was keeping the change from coming into the site, but it should work now.

**Comment by Buteor** [comment]
what is flair example on spreedsheet for flair_id and flair_text?

**Comment by tjordanb** [comment]
I don't think this currently works yet, at least not on the cronnit.com site (custom hosted ones may work for all I know)

I am in no way fluent in any code, but based on what @lerokko stated above, it appears that line of code is meant to convert the flair_identifier tag into the flair_id tag when it's actually posted, but instead flair_identifier has replaced both instances of flair_id and therefore gets ignored by Reddit?  Does that seem right?

@krisives could you ever take a look at this again & potentially update the .csv template to include the fields it needs?
